pkg/controller/ttlafterfinished/util.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ttlafterfinished
+
+import (
+	"time"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func NewJobsDelayCleanupQueue() *cache.Heap {
+	keyFn := cache.DeletionHandlingMetaNamespaceKeyFunc
+	lessFn := soonerExpiringJob
+	return cache.NewHeap(keyFn, lessFn)
+}
+
+func soonerExpiringJob(finishedJob1, finishedJob2 interface{}) bool {
+	j1, ok := finishedJob1.(*batch.Job)
+	if !ok {
+		return true
```

Comment:
I believe this is a bug. It should be true that soonerExpiringJob(f1, f2) != soonerExpiringJob(f2, f1) as long as f1 != f2. This statement will violate that. (i.e. if both j1 and j2 are not batch jobs you should still be consistent about their sort order. You could use memory address or something else to achieve that.

Reply:
> It should be true that soonerExpiringJob(f1, f2) != soonerExpiringJob(f2, f1) as long as f1 != f2.

Is it required? If `f1 != f2` and `f1` and `f2` expire at the same time, both should return `false`. In [sort/sort.go](https://golang.org/src/sort/sort.go) it's also implemented similarly:
```go 
func (p StringSlice) Less(i, j int) bool { return p[i] < p[j] }
```

I want the error cases (i.e. non-Job objects or unfinished objects) to be placed on the top of the heap so that it can be dropped earlier instead of occupying space. It doesn't matter which error-case object is popped before the other one.

Reply:
So comparison could be achieved by reflect.ValueOf(f1).Pointer() < reflect.ValueOf(f2).Pointer().
We may just want to consider that f1 == f2 for purposes of this sort.
That having been said if both f1 and f2 are not batch jobs we should return false (ie they are equal not less) but right now you will return true.

Reply:
Updated the code to treat error cases are f1==f2.